src/subsidy/subsidy.service.ts

```diff
@@ -0,0 +1,105 @@
+import { ModifyAllowanceDto } from '~/subsidy/dto/modify-allowance.dto';
+import { QuitSubsidyDto } from '~/subsidy/dto/quit-subsidy.dto';
+import { TransactionsService } from '~/transactions/transactions.service';
+
+@Injectable()
+export class SubsidyService {
+  constructor(
+    private readonly polymeshService: PolymeshService,
+    private readonly transactionsService: TransactionsService,
+    private readonly accountsService: AccountsService
+  ) {}
+
+  public async getSubsidy(address: string): Promise<SubsidyWithAllowance | null> {
+    const account = await this.accountsService.findOne(address);
+    return account.getSubsidy();
+  }
+
+  public findOne(beneficiary: string, subsidizer: string): Subsidy {
+    return this.polymeshService.polymeshApi.accountManagement.getSubsidy({
+      beneficiary,
+      subsidizer,
+    });
+  }
+
+  public async subsidizeAccount(params: CreateSubsidyDto): ServiceReturn<AuthorizationRequest> {
+    const { signer, webhookUrl, beneficiary, allowance } = params;
+
+    const { subsidizeAccount } = this.polymeshService.polymeshApi.accountManagement;
+
+    return this.transactionsService.submit(
+      subsidizeAccount,
+      { beneficiary, allowance },
+      { signer, webhookUrl }
+    );
+  }
+
+  public async quit(params: QuitSubsidyDto): ServiceReturn<void> {
+    const { signer, webhookUrl, beneficiary, subsidizer } = params;
+
+    const address = await this.transactionsService.getSigningAccount(signer);
+
+    let subsidy: Subsidy;
+    if (beneficiary) {
+      subsidy = this.findOne(beneficiary, address);
+    } else if (subsidizer) {
+      subsidy = this.findOne(address, subsidizer);
+    } else {
+      throw new BadRequestException('Beneficiary or subsidizer should be specified');
+    }
+
+    return this.transactionsService.submit(subsidy.quit, {}, { signer, webhookUrl });
+  }
+
+  public async modifyAllowance(params: ModifyAllowanceDto): ServiceReturn<void> {
+    const { signer, webhookUrl, beneficiary, allowance, operation } = params;
+
+    const address = await this.transactionsService.getSigningAccount(signer);
+
+    const subsidy = this.findOne(beneficiary, address);
+
+    const procedureMap = {
+      [AllowanceOperation.Set]: subsidy.setAllowance,
+      [AllowanceOperation.Increase]: subsidy.increaseAllowance,
+      [AllowanceOperation.Decrease]: subsidy.decreaseAllowance,
+    };
+
+    return this.transactionsService.submit(
+      procedureMap[operation],
+      { allowance },
+      { signer, webhookUrl }
+    );
+  }
+
+  public async getAllowance(beneficiary: string, subsidizer: string): Promise<BigNumber> {
+    const subsidy = this.findOne(beneficiary, subsidizer);
+
+    try {
+      return await subsidy.getAllowance();
+    } catch (err: unknown) {
+      if (isPolymeshError(err)) {
+        const { code } = err;
+        if (code === ErrorCode.DataUnavailable) {
+          throw new NotFoundException('The Subsidy no longer exists');
```

Comment:
I think its better to handle all the SDK errors in one place with the `handleSdkError` function. We can update that method to accept `unknown` instead of `Error` to avoid needing the `as Error` cast for the caller.

The SDK message should be descriptive enough to return directly, and if its not we should update that message.

To me this looks messy and is a lot of lines to repeat. The SDK errors should be good enough to be handled in.a generic manner imo.

That said I know this pattern exists already and won't block the PR for this.

Reply:
I agree with you. I'll change it here and I've created a separate task - [DA-468](https://polymesh.atlassian.net/browse/DA-468) to replace all other instances and cleanup the test cases for the same.